### PR TITLE
Parser improvements: split ident rule into un/-quoted ident rules and use AST visitors for dataType child nodes.

### PR DESCRIPTION
**Pull request by kovrus**
## Checklist

 - [ ] User relevant changes are recorded in ``CHANGES.txt``
 - [x] Touched code is covered by tests
 - [ ] Documentation has been updated if necessary
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed

### CODE REVIEW

File: sql-parser/src/main/antlr/SqlBase.g4 [hunk 1]
```diff
@@ -529,12 +533,10 @@ rerouteOption
     ;
 
 dataType
-    : IDENTIFIER
-    | quotedIdentifier
-    | nonReserved
-    | objectTypeDefinition
-    | arrayTypeDefinition
-    | setTypeDefinition
+    : ident                     #dataTypeIdent
```
Comment (kovrus) [inline sql-parser/src/main/antlr/SqlBase.g4]: replacing `quotedIdentifier`, `nonReserved`, `IDENTIFIER` with ident results in exactly the same behavior.

File: sql-parser/src/main/antlr/SqlBase.g4 [hunk 2]
```diff
@@ -371,16 +371,20 @@ idents
...
-    | quotedIdentifier                                                               #quotedIdentifierAlternative
-    | nonReserved                                                                    #unquotedIdentifier
-    | BACKQUOTED_IDENTIFIER                                                          #backQuotedIdentifier
-    | DIGIT_IDENTIFIER                                                               #digitIdentifier
-    | COLON_IDENT                                                                    #colonIdentifier
+    : unquotedIdent
+    | quotedIdent
     ;
 
-quotedIdentifier
```
Comment (kovrus) [inline sql-parser/src/main/antlr/SqlBase.g4]: it was needed to remove quotes with `exitQuotedIdentifier`. we can do it in the `quotedIdentifier` visitor as well.

**Comment by mfussenegger** [comment]
What's the motivation behind this change?

**Comment by kovrus** [comment]
@mfussenegger The reason is #8441. The data type visitor starts to grow and has quite few conditions https://github.com/crate/crate/pull/8441/files#diff-f5b7cb3fa46cea9783b05e7068cd45e0R1601

**Pull request merged**